apps/shared/avifutil.c

```diff
@@ -220,3 +220,74 @@ void avifDumpDiagnostics(const avifDiagnostics * diag)
     printf("Diagnostics:\n");
     printf(" * %s\n", diag->error);
 }
+
+// ---------------------------------------------------------------------------
+// avifQueryMaxThreads (separated into OS implementations)
+
+#if defined(_WIN32)
+
+// Windows
+
+#pragma warning(disable : 5031)
+#pragma warning(disable : 5032)
+#include <windows.h>
+
+int avifQueryMaxThreads(void)
+{
+    int numCPU;
+    SYSTEM_INFO sysinfo;
+    GetSystemInfo(&sysinfo);
+    numCPU = sysinfo.dwNumberOfProcessors;
+    return numCPU;
+}
+
+#elif defined(__APPLE__)
+
+// Apple
+
+#include <sys/sysctl.h>
+
+int avifQueryMaxThreads()
+{
+    int mib[4];
+    int numCPU;
+    size_t len = sizeof(numCPU);
+
+    /* set the mib for hw.ncpu */
+    mib[0] = CTL_HW;
+    mib[1] = HW_AVAILCPU; // alternatively, try HW_NCPU;
+
+    /* get the number of CPUs from the system */
+    sysctl(mib, 2, &numCPU, &len, NULL, 0);
+
+    if (numCPU < 1) {
+        mib[1] = HW_NCPU;
+        sysctl(mib, 2, &numCPU, &len, NULL, 0);
+        if (numCPU < 1)
+            numCPU = 1;
+    }
+    return numCPU;
+}
+
+#elif defined(__EMSCRIPTEN__)
+
+// Emscripten
+
+int avifQueryMaxThreads()
+{
+    return 1;
+}
+
+#else
+
+// POSIX
+
+#include <unistd.h>
+
+int avifQueryMaxThreads()
+{
+    int numCPU = (int)sysconf(_SC_NPROCESSORS_ONLN);
```

Comment:
I vaguely remember something about BSD not having `_SC_NPROCESSORS_ONLN`, but don't remember what it used instead

Reply:
This is basically just a recontribution of my original block of code from `colorist`. I'm perfectly willing to adjust it to support more platforms as we encounter issues.

Reply:
Wait, nvm, it seems I had it backwards, it was that `_SC_NPROCESSORS_ONLN` was available on BSD, but some other options weren't available

Reply:
This kind of code is available in a lot of open-source libraries, so it is easy to find sample code. Some of the platform-specific code may no longer be necessary for the current versions of those Unix variants because they generally try to conform to POSIX or the Single Unix Specification or be compatible with Linux. For example, the code for __APPLE__ above may no longer be necessary for current versions of macOS and iOS.

A good strategy is to start with the POSIX code here and add platform-specific code only in response to bug report.

Re: BSD: The Chromium source code only has [a special case for OpenBSD](https://source.chromium.org/chromium/chromium/src/+/main:base/system/sys_info_openbsd.cc) in its POSIX code, which is similar to your __APPLE__ code above:

```
// static
int SysInfo::NumberOfProcessors() {
  int mib[] = {CTL_HW, HW_NCPU};
  int ncpu;
  size_t size = sizeof(ncpu);
  if (sysctl(mib, base::size(mib), &ncpu, &size, NULL, 0) < 0) {
    NOTREACHED();
    return 1;
  }
  return ncpu;
}
```

Chromium calls `sysconf(_SC_NPROCESSORS_CONF)` for the [general POSIX case](https://source.chromium.org/chromium/chromium/src/+/main:base/system/sys_info_posix.cc). Whether `_SC_NPROCESSORS_CONF` or `_SC_NPROCESSORS_ONLN` is more appropriate doesn't really matter in practice.